osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/DeviceKeyProcessingService.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2021 Alliander N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ */
+package org.opensmartgridplatform.adapter.protocol.dlms.application.services;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Date;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.entities.DlmsDevice;
+import org.opensmartgridplatform.adapter.protocol.dlms.domain.repositories.DlmsDeviceRepository;
+import org.opensmartgridplatform.adapter.protocol.dlms.exceptions.DeviceKeyProcessAlreadyRunningException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DeviceKeyProcessingService {
+
+  @Value("#{T(java.time.Duration).parse('${device.key.processing.timeout:PT5M}')}")
+  private Duration deviceKeyProcessingTimeout;
+
+  private final DlmsDeviceRepository dlmsDeviceRepository;
+
+  @Autowired
+  public DeviceKeyProcessingService(final DlmsDeviceRepository dlmsDeviceRepository) {
+    this.dlmsDeviceRepository = dlmsDeviceRepository;
+  }
+
+  public void startProcessing(final String deviceIdentification)
+      throws DeviceKeyProcessAlreadyRunningException {
+
+    final DlmsDevice dlmsDevice =
+        this.dlmsDeviceRepository.findByDeviceIdentification(deviceIdentification);
+    final Date keyProcessingStartTime = dlmsDevice.getKeyProcessingStartTime();
+    final Instant timeoutTime = Instant.now().minus(this.deviceKeyProcessingTimeout);
+    final boolean processIsAlreadyRunning =
+        keyProcessingStartTime != null && timeoutTime.isBefore(keyProcessingStartTime.toInstant());
+    if (processIsAlreadyRunning) {
```

Comment:
What would happen if two or more processes were running for the same device and got to the point where `startProcessing` is called, entered this method and more than one determined `processIsAlreadyRunning` to be false, before saving the `keyProcessingStartTime`?
It looks like in such a situation, this would not protect against concurrent device key operations like it should.

I suspect better protection would come for example from an update query in the repository which sets the `keyProcessingStartTime` for the device identifiied by `deviceIdentification` under the condition its current value is NULL or is older than what is now called the `timeoutTime`. If this updates zero records, then a process was already running, if it updates a single record than the current process may continue its business.

Reply:
Logic in DeviceKeyProcessingService
```
    final Instant oldestStartTimeNotConsiderTimedOut =
        Instant.now().minus(this.deviceKeyProcessingTimeout);
    final int updatedRecords =
        this.dlmsDeviceRepository.setProcessingStartTime(
            dlmsDevice.getDeviceIdentification(), oldestStartTimeNotConsiderTimedOut);
    if (updatedRecords == 0) {
      throw new DeviceKeyProcessAlreadyRunningException();
    }
```

Update query in DlmsDeviceRepository
```
  @Modifying
  @Query(
      value =
          "UPDATE DlmsDevice"
              + "   SET keyProcessingStartTime = CURRENT_TIMESTAMP"
              + " WHERE deviceIdentification = :deviceIdentification"
              + "   AND (keyProcessingStartTime IS NULL OR"
              + "        keyProcessingStartTime < :oldestStartTimeNotConsiderTimedOut)")
  int setProcessingStartTime(
      @Param("deviceIdentification") String deviceIdentification,
      @Param("oldestStartTimeNotConsiderTimedOut") Instant oldestStartTimeNotConsiderTimedOut);

```